src/engine/source/metrics/src/metrics.cpp

```diff
@@ -3,15 +3,477 @@
+        {
+            m_doubleCounter.at(counterName)->Add(value);
+        }
+    }
+    else
+    {
+        throw std::runtime_error {"The Counter" + counterName + " has not been created."};
+    }
+}
+
+void Metrics::addCounterValue(std::string counterName, const uint64_t value) const
+{
+    if (m_uint64Counter.find(counterName) != m_uint64Counter.end())
+    {
+        if(controller.at(counterName))
+        {
+            m_uint64Counter.at(counterName)->Add(value);
+        }
+    }
+    else
+    {
+        throw std::runtime_error {"The Counter" + counterName + " has not been created."};
+    }
+}
+
+void Metrics::initHistogram(const std::string& name, SubType subType)
+{
+    auto histogramName = name + "_histogram";
+    auto provider = opentelemetry::metrics::Provider::GetMeterProvider();
+    opentelemetry::nostd::shared_ptr<opentelemetry::metrics::Meter> meter = provider->GetMeter(name);
+
+    switch (subType)
+    {
+        case SubType::Double:
+        {
+            if (m_doubleHistogram.find(name) == m_doubleHistogram.end())
+            {
+                m_doubleHistogram.insert({name, meter->CreateDoubleHistogram(histogramName)});
+            }
+            else
+            {
+                throw std::runtime_error {"The Histogram " + name + " has already been created."};
+            }
+            break;
+        }
+        case SubType::Int64:
+        {
+            throw std::runtime_error {"Histogram type instrument does not accept integers."};
+        }
+        case SubType::UInt64:
+        {
+            if (m_uint64Histogram.find(name) == m_uint64Histogram.end())
+            {
+                m_uint64Histogram.insert({name, meter->CreateUInt64Histogram(histogramName)});
+            }
+            else
+            {
+                throw std::runtime_error {"The Histogram " + name + " has already been created."};
+            }
+        }
+    }
+
+    m_context = opentelemetry::context::Context{};
+}
+
+void Metrics::addHistogramValue(std::string histogramName, const double value) const
+{
+    if (m_doubleHistogram.find(histogramName) != m_doubleHistogram.end())
+    {
+        if(controller.at(histogramName))
+        {
+            std::map<std::string, std::string> labels;
+            auto labelkv = opentelemetry::common::KeyValueIterableView<decltype(labels)>{labels};
+            m_doubleHistogram.at(histogramName)->Record(value, labelkv, m_context);
+        }
+    }
+    else
+    {
+        throw std::runtime_error {"The Histogram" + histogramName + " has not been created."};
+    }
+}
+
+void Metrics::addHistogramValue(std::string histogramName, const uint64_t value) const
+{
+    if (m_uint64Histogram.find(histogramName) != m_uint64Histogram.end())
+    {
+        if(controller.at(histogramName))
+        {
+            std::map<std::string, std::string> labels;
+            auto labelkv = opentelemetry::common::KeyValueIterableView<decltype(labels)>{labels};
```

Comment:
I think those labels are not used, you could directly use:

`m_uint64Histogram.at(histogramName)->Record(value, m_context);`